### DISCUSSION TITLE
Build requirements and editable install

### DISCUSSION THREAD

**Discussion by mancellin** [mesonbuild/meson-python#504] [Q&A] [answered]
Hi,

Thank you for this package, I'm a happy user of meson-python and I have a quick question to better understand how it works (which actually might be a PEP 517 question as much as a meson-python question).

When listing the package build requirements in `pyproject.toml`, Meson-python's documentation suggests
```toml
[build-system]
build-backend = 'mesonpy'
requires = ['meson-python', 'cython']
```
without including `meson` itself or `ninja`. And indeed for my project I never had to list `meson` or `ninja` there.

On the other hand, when doing an editable install, the documentation suggests
```shell
python -m pip install meson-python meson ninja
```
And omitting `ninja` indeed leads to a failure of the installation (https://github.com/capytaine/capytaine/issues/355#issuecomment-1731795990)

Is there something different happening when requiring `meson-python` for the isolated build environment and pip-installing it myself? Why do I need to list explicitly ninja in one case and not the other?

**Comment by rgommers** [comment, marked as answer]
Hi @mancellin, thanks for the question, and glad you're happy so far:)

The reason is that editable installs use `--no-build-isolation`, which means that any necessary build dependencies must already be installed in the target environment, and `pip` + `meson-python` will not attempt to install them when you call `pip install -e . --no-build-isolation`. I hope that explains it.

Other note: `ninja` isn't really a Python package, it just happens to be installable from PyPI but it's easier/nicer to install it with your system package manager once and have it always available.

**Reply by mancellin** [reply]
Thank you for your answer.

By reading a bit more (e.g. #485), I think I found what was bugging me:
There is a mechanism with which the build backend (e.g. `meson-python`) can add to the build environment packages that where not specified in `pyproject.toml`.
Hence the list of packages that I have to install manually for the editable install might not match the one of the `build-system.requires` entry of `pyproject.toml`.

**Reply by rgommers** [reply]
Yes, that is right - and indeed a little annoying. It's an unfortunate downside from PyPI being (mostly) Python-packages-only and needing a system package manager for non-Python dependencies like compilers and C/C++ libraries and tools. `ninja` should normally come from the system package manager, so that's checked for first - and then it's inserted if and only if there's no system install of it detected.
